### PR TITLE
add digitize/bucketize

### PR DESCRIPTION
**Pull request by mtar**
## Description

implementation of digitize and bucketize

Issue/s resolved: #926

## Changes proposed:
- add bucketize()
- add digitize()

## Type of change

-New Feature

## Due Diligence
- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

File: heat/core/statistics.py [hunk 1]
```diff
@@ -388,6 +390,79 @@ def bincount(x: DNDarray, weights: Optional[DNDarray] = None, minlength: int = 0
...
+    >>> ht.bucketize(v, boundaries)
+    DNDarray([[1, 3, 4],
+              [1, 3, 4]], dtype=ht.int64, device=cpu:0, split=None)
+    >>> ht.bucketize(v, boundaries, right=True)
+    DNDarray([[2, 3, 5],
+              [2, 3, 5]], dtype=ht.int64, device=cpu:0, split=None)
+    """
+    if isinstance(boundaries, DNDarray):
+        if boundaries.is_distributed():
+            raise RuntimeError("'boundaries' must be undistributed.")
```
Comment (mtar) [inline heat/core/statistics.py:450]:
```suggestion
            raise RuntimeError("'boundaries' must not be distributed.")
```

File: heat/core/statistics.py [hunk 2]
```diff
@@ -463,6 +538,81 @@ def cov(
...
+    >>> x = ht.array([1.2, 10.0, 12.4, 15.5, 20.])
+    >>> bins = ht.array([0, 5, 10, 15, 20])
+    >>> ht.digitize(x,bins,right=True)
+    DNDarray([1, 2, 3, 4, 4], dtype=ht.int64, device=cpu:0, split=None)
+    >>> ht.digitize(x,bins,right=False)
+    DNDarray([1, 3, 3, 4, 5], dtype=ht.int64, device=cpu:0, split=None)
+    """
+    if isinstance(bins, DNDarray):
+        if bins.is_distributed():
+            raise RuntimeError("'bins' must be undistributed.")
```
Comment (mtar) [inline heat/core/statistics.py:589]:
```suggestion
            raise RuntimeError("'bins' must not be distributed.")
```

**Review by ClaudiaComito — COMMENTED** [review]
@mtar thanks for this. I'm ready to approve it, maybe mention in the documentation that `bins` and `boundaries` cannot be distributed?

File: heat/core/statistics.py [hunk 3]
```diff
@@ -388,6 +390,79 @@ def bincount(x: DNDarray, weights: Optional[DNDarray] = None, minlength: int = 0
...
+) -> DNDarray:
+    """
+    Returns the indices of the buckets to which each value in the input belongs, where the boundaries of the buckets are set by boundaries.
+
+    Parameters
+    ----------
+    input : DNDarray
+        The input array.
+    boundaries : DNDarray or torch.Tensor
+        monotonically increasing sequence defining the bucket boundaries, 1-dimensional
```
Comment (mtar) [inline heat/core/statistics.py:408]:
```suggestion
        monotonically increasing sequence defining the bucket boundaries, 1-dimensional, not distributed
```

File: heat/core/statistics.py [hunk 4]
```diff
@@ -463,6 +538,81 @@ def cov(
...
+    """
+    Return the indices of the bins to which each value in the input array `x` belongs.
+    If values in `x` are beyond the bounds of bins, 0 or len(bins) is returned as appropriate.
+
+    Parameters
+    ----------
+    x : DNDarray
+        The input array
+    bins : DNDarray or torch.Tensor
+        A 1-dimensional array containing a monotonic sequence describing the bin boundaries.
```
Comment (ClaudiaComito) [inline heat/core/statistics.py:551]: cannot be distributed

**Review by ClaudiaComito — APPROVED** [review]
Thanks @mtar !

**Pull request merged**